prow/plugins/jira/jira.go

```diff
@@ -0,0 +1,201 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jira
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/andygrunwald/go-jira"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+	jiraclient "k8s.io/test-infra/prow/jira"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	PluginName = "jira"
+)
+
+var (
+	issueNameRegex = regexp.MustCompile(`\b[a-zA-Z]+-[0-9]+\b`)
+)
+
+func init() {
+	plugins.RegisterGenericCommentHandler(PluginName, handleGenericComment, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, _ []config.OrgRepo) (*pluginhelp.PluginHelp, error) {
+	// The Config field is omitted because this plugin is not configurable.
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The Jira plugin links Pull Requests and Issues to Jira issues",
+	}
+	return pluginHelp, nil
+}
+
+type githubClient interface {
+	EditComment(org, repo string, id int, comment string) error
+	GetIssue(org, repo string, number int) (*github.Issue, error)
+	EditIssue(org, repo string, number int, issue *github.Issue) (*github.Issue, error)
+}
+
+func handleGenericComment(pc plugins.Agent, e github.GenericCommentEvent) error {
+	return handle(pc.JiraClient, pc.GitHubClient, pc.Logger, &e)
+}
+
+func handle(jc jiraclient.Client, ghc githubClient, log *logrus.Entry, e *github.GenericCommentEvent) error {
+	// Nothing to do on deletion
+	if e.Action == github.GenericCommentActionDeleted {
+		return nil
+	}
+
+	issueCandidateNames := issueNameRegex.FindAllString(e.Body, -1)
+	issueCandidateNames = append(issueCandidateNames, issueNameRegex.FindAllString(e.IssueTitle, -1)...)
+	if len(issueCandidateNames) == 0 {
+		return nil
+	}
+
+	referencedIssues := sets.String{}
+	for _, match := range issueCandidateNames {
+		if referencedIssues.Has(match) {
+			continue
+		}
+		_, err := jc.GetIssue(match)
+		if err != nil {
+			if !jiraclient.IsNotFound(err) {
+				log.WithError(err).WithField("Issue", match).Error("Failed to get issue")
+			}
+			continue
+		}
+		referencedIssues.Insert(match)
+	}
+
+	wg := &sync.WaitGroup{}
+	for _, issue := range referencedIssues.List() {
+		wg.Add(1)
+		go func(issue string) {
+			defer wg.Done()
+			if err := upsertGithubLinkToIssue(log, issue, jc, e); err != nil {
+				log.WithField("Issue", issue).WithError(err).Error("Failed to ensure Github link on Jira issue")
+			}
+		}(issue)
+	}
+
+	if err := updateComment(e, referencedIssues.UnsortedList(), jc.JiraURL(), ghc); err != nil {
+		logrus.WithError(err).Error("Failed to insert links into body")
+	}
+	wg.Wait()
+
+	return nil
+}
+
+func updateComment(e *github.GenericCommentEvent, validIssues []string, jiraBaseURL string, ghc githubClient) error {
```

Comment:
This is new behavior for Prow - editing user's comments - what is wrong with responding with our own comment from the bot that we added a link?

Reply:
Well, because its a lot more practical. Someone uses a link in their comment, you can just click on it. Thats a lot nicer than having to scroll to find some bot message. The bot message makes a lot of sense for for example approval and tests, because the message for those always belongs at the end, but this isn't really the case here. You can see it btw in action here https://github.com/alvaro-bot/test-prow/pull/2 and https://github.com/alvaro-bot/test-prow/issues/3